module/os/freebsd/zfs/zfs_vnops_os.c

```diff
@@ -6333,3 +6338,356 @@ VFS_VOP_VECTOR_REGISTER(zfs_shareops);
+		error = EINVAL;
+		goto out;
+	}
+
+	/*
+	 * Avoid lock order reversal.
+	 */
+	if (cbl->srcvp < cbl->dstvp) {
+		error = clonefile_lock_src(cbl->srcfp, cbl->srcvp, td);
+		if (error == 0) {
+			cbl->srcvnlock = B_TRUE;
+			error = clonefile_lock_dst(cbl->dstfp, cbl->dstvp, td,
+			    &cbl->dstmp);
+			if (error == 0) {
+				cbl->dstvnlock = B_TRUE;
+			}
+		}
+	} else {
+		error = clonefile_lock_dst(cbl->dstfp, cbl->dstvp, td,
+		    &cbl->dstmp);
+		if (error == 0) {
+			cbl->dstvnlock = B_TRUE;
+			if (cbl->srcvp != cbl->dstvp) {
+				error = clonefile_lock_src(cbl->srcfp,
+				    cbl->srcvp, td);
+				if (error == 0) {
+					cbl->srcvnlock = B_TRUE;
+				}
+			}
+		}
+	}
+	if (error != 0) {
+		goto out;
+	}
+
+	cbl->srcioflag = clonefile_get_ioflag(cbl->srcfp);
+	cbl->dstioflag = clonefile_get_ioflag(cbl->dstfp);
+
+out:
+	if (error != 0) {
+		clonefile_unlock(cbl, td);
+	}
+
+	return (error);
+}
+
+static int
+kern_fclonerange(struct thread *td, int srcfd, uint64_t srcoffset,
+    uint64_t length, int dstfd, uint64_t dstoffset, ssize_t *donep)
+{
+	struct clonefile_lock cbl;
+	ssize_t done;
+	int error;
+
+	*donep = 0;
+
+	error = clonefile_lock(srcfd, dstfd, td, &cbl);
+	if (error != 0) {
+		return (error);
+	}
+
+	/*
+	 * Check if both source and destination vnodes are on a ZFS file system.
+	 */
+	error = zfs_verify_vp(cbl.srcvp);
+	if (error == 0) {
+		error = zfs_verify_vp(cbl.dstvp);
+	}
+	if (error != 0) {
+		clonefile_unlock(&cbl, td);
+		return (error);
+	}
+
+	error = zfs_clone_range(VTOZ(cbl.srcvp), srcoffset, length,
+	    ioflags(cbl.srcioflag), VTOZ(cbl.dstvp), dstoffset,
+	    ioflags(cbl.dstioflag), td->td_ucred, &done);
+
+	clonefile_unlock(&cbl, td);
+
+	*donep = done;
+
+	return (done > 0 ? 0 : error);
+}
+
+struct fclonefile_args {
+	int64_t	srcfd;
+	int64_t	dstfd;
+};
+
+/*
+ * fclonefile(2) syscall implementation.
+ */
+static int
+sys_fclonefile(struct thread *td, void *arg)
+{
+	struct fclonefile_args *args;
+	ssize_t done;
+	int error;
+
+	args = arg;
+
+	error = kern_fclonerange(td, (int)args->srcfd, 0ULL, UINT64_MAX,
+	    (int)args->dstfd, 0ULL, &done);
+
+	td->td_retval[0] = done;
+
+	return (done > 0 ? 0 : error);
+}
+
+/*
+ * The `sysent' for the new syscall
+ */
+static struct sysent fclonefile_sysent = {
+	.sy_narg = 4,
+	.sy_call = sys_fclonefile
+};
+
+/*
+ * The offset in sysent where the syscall is allocated.
+ */
+static int fclonefile_num = NO_SYSCALL;
+
+struct fclonerange_args {
+	int64_t	srcfd;
+	uint64_t srcoffset;
+	uint64_t length;
+	int64_t	dstfd;
+	uint64_t dstoffset;
+};
+
+/*
+ * fclonerange(2) syscall implementation.
+ */
+static int
+sys_fclonerange(struct thread *td, void *arg)
+{
+	struct fclonerange_args *args;
+	ssize_t done;
+	int error;
+
+	args = arg;
+
+	error = kern_fclonerange(td, (int)args->srcfd, args->srcoffset,
+	    args->length, (int)args->dstfd, args->dstoffset, &done);
+
+	td->td_retval[0] = done;
+
+	return (done > 0 ? 0 : error);
+}
```

Comment:
As I see, FreeBSD already has sys_copy_file_range()/kern_copy_file_range() and even linux_copy_file_range().  Aren't you duplicating?

Reply:
To be honest I was under impression that copy_file_range(2) always copied the data, just avoiding copying the data to and from userland, but Linux's manual page does mention that reflinks can be used. So, yes, that is a good candidate to use. I'd still prefer to have a flag controlling this behavior, but on Linux, I guess, reflink is used when available.

All in all, please ignore the syscalls for now. This will be done separately, maybe under sponsorship of the FreeBSD Foundation.